getnacelle/nacelle-react
# [packages/react-hooks] Added Flexibility for Cart Storage

**Story:** [DD-428](https://nacelle.atlassian.net/browse/DD-428)

> As a merchant developer, I want to be able to customize @nacelle/react-hooks' use of the Storage API, so that I have more control of my customer's cart persistence

### Type of Change

- [x] Breaking feature (requires bump to next major version)

### What is being changed and why?

#### Allow Custom Storage Key

_Non-Breaking Change_

When persisting the cart in browser storage, the key `'cart'` was used, without any ability to control the key name. This can cause headaches when migrating from `@nacelle/react-hooks@3.x`, which formatted `CartItems`s differently.

This change allows merchant developers to upgrade from  `@nacelle/react-hooks@3.x` without creating any collisions between mis-matched carts in `localStorage`, for example, when customers who have an older cart saved in their `localStorage`.

##### Demonstration

https://user-images.githubusercontent.com/5732000/117478517-43c05780-af2d-11eb-8c7f-90d5739eb6d2.mov

#### New Session Storage Option

_Breaking Change_

Allows the merchant developer to persist the cart with [Session Storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) instead of Local Storage. The breaking change comes from the removal of the `<CartProvider />`'s `useLocalStorage` prop, in favor of a `storage` prop, which can receive `'local'`, `'session'`, or `null`.



##### Demonstration

https://user-images.githubusercontent.com/5732000/117479207-0a3c1c00-af2e-11eb-8a20-7ebb98110856.mov

### How to Test



From the monorepo root:

1. Unit Tests
- `npm run bootstrap`
- `cd packges/react-hooks && npm run test:ci`

2. Example Projects
- go to either `examples/nextjs` or `examples/gatsby` and `npm run dev`
- in the App's root (either `examples/nextjs/pages/_app.js` or `examples/gatsby/gatsby-browser.js`), experiment with the `<CartProvider />`'s `cacheKey` and `storage` props
- add some items to the cart, remove them, update their quantity, etc. and verify that changes are made in either Local Storage or Session Storage
